src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -271,44 +302,66 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
             }
             walkerBuilder.Free();
 
-            // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
-            for (int i = 0; i < state.Generators.Length; i++)
+            // PROTOTYPE(source-generators): we don't need to run at all if none of the inputs have changes.
+            var driverStateBuilder = new DriverStateTable.Builder(_state.StateTable, cancellationToken);
+            driverStateBuilder.SetInputState(SharedInputNodes.Compilation, NodeStateTable<Compilation>.WithSingleItem(compilation, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.AnalzerConfigOptions, NodeStateTable<AnalyzerConfigOptionsProvider>.WithSingleItem(_state.OptionsProvider, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.ParseOptions, NodeStateTable<ParseOptions>.WithSingleItem(_state.ParseOptions, EntryState.Added));
+
+            var additionalBuilder = new NodeStateTable<AdditionalText>.Builder(_state.AdditionalTexts.Length);
+            foreach (var text in _state.AdditionalTexts)
             {
-                var generator = state.Generators[i];
-                var generatorState = stateBuilder[i];
+                additionalBuilder.AddEntries(ImmutableArray.Create(text), EntryState.Added);
+            }
+            driverStateBuilder.SetInputState(SharedInputNodes.AdditionalTexts, additionalBuilder.ToImmutableAndFree());
```

Comment:
Had a comment about de-duplicating the logic between this and `AddAdditionalTexts` above. Didn't see that addressed here (PROTTYPE would be fine)